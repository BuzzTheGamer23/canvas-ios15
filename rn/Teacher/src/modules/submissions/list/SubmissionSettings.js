//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import React, { PureComponent } from 'react'
import {
  View,
} from 'react-native'
import Screen from '../../../routing/Screen'
import i18n from 'format-message'
import RowWithSwitch from '../../../common/components/rows/RowWithSwitch'
import { connect } from 'react-redux'
import AssignmentActions from '../../assignments/actions'
import branding from '../../../common/branding'
import { SubTitle } from '../../../common/text'

type SubmissionSettingsOwnProps = {
  courseID: string,
  assignmentID: string,
  navigator: Navigator,
}
type SubmissionSettingsDataProps = {
  anonymous: boolean,
  muted: boolean,
  assignment: Assignment,
}
type SubmissionSettingsActions = {
  anonymousGrading: (string, string, boolean) => void,
  updateAssignment: (string, Assignment, Assignment) => void,
}

type SubmissionSettingsProps =
  SubmissionSettingsOwnProps &
  SubmissionSettingsDataProps &
  SubmissionSettingsActions

export class SubmissionSettings extends PureComponent {
  props: SubmissionSettingsProps

  dismiss = () => {
    this.props.navigator.dismiss()
  }

  toggleAnonymousGrading = (value: boolean) => {
    this.props.anonymousGrading(
      this.props.courseID,
      this.props.assignmentID,
      value
    )
  }

  toggleMutedGrading = (value: boolean) => {
    this.props.updateAssignment(
      this.props.courseID,
      {
        ...this.props.assignment,
        muted: value,
      },
      this.props.assignment
    )
  }

  render () {
    return (
      <Screen
        title={i18n('Submission Settings')}
        navBarButtonColor={branding.primaryButtonColor}
        rightBarButtons={[{
          title: i18n('Done'),
          style: 'done',
          testID: 'submission-settings.done',
          action: this.dismiss,
        }]}
      >
        <View>
          <RowWithSwitch
            border='bottom'
            height={60}
            title={i18n('Mute Grades')}
            value={this.props.muted}
            onValueChange={this.toggleMutedGrading}
            identifier='submission-settings.muted'
          />
          <RowWithSwitch
            border='bottom'
            height={60}
            title={i18n('Anonymous Grading')}
            value={this.props.anonymous}
            onValueChange={this.toggleAnonymousGrading}
            identifier='submission-settings.anonymous'
          />
          <SubTitle style={{ paddingHorizontal: 12, paddingVertical: 4 }}>
            {i18n('This will anonymize each student and shuffle the submission list.')}
          </SubTitle>
        </View>
      </Screen>
    )
  }
}

export function mapStateToProps (state: AppState, ownProps: SubmissionSettingsOwnProps): SubmissionSettingsDataProps {
  let anonymous = !!state.entities.assignments[ownProps.assignmentID].anonymousGradingOn
  let assignment = state.entities.assignments[ownProps.assignmentID].data
  let muted = !!assignment.muted
  return { anonymous, muted, assignment }
}
const Connect = connect(mapStateToProps, AssignmentActions)(SubmissionSettings)
export default (Connect: any)

