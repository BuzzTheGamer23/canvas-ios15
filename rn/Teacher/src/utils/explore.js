/* @flow */

function hasProp (object: any, key: string, value: any): boolean {
  if (object instanceof Object) {
    const props = object.props
    if (props && props instanceof Object) {
      for (var prop in props) {
        if (prop === key && props[prop] === value) {
          return true
        }
      }
    }
  }

  return false
}

function selectByProp (object: any, propKey: string, value: any): ?any {
  if (!(typeof object === 'object')) {
    return null
  }

  if (hasProp(object, propKey, value)) {
    return object
  }

  let result = null

  for (var key in object) {
    if (object[key] instanceof Array) {
      object[key].forEach((e) => {
        result = selectByProp(e, propKey, value)
        if (result) {
          return result
        }
      })
    }

    result = selectByProp(object[key], propKey, value)
    if (result) {
      return result
    }
  }

  return result
}

class ComponentExplorer {
  component: any

  constructor (component: any) {
    this.component = component
  }

  selectByProp (propKey: string, value: any): ?any {
    return selectByProp(this.component, propKey, value)
  }

  selectByID (id: string): ?any {
    return this.selectByProp('testID', id)
  }
}

export default function explore (component: any): ComponentExplorer {
  return new ComponentExplorer(component)
}
