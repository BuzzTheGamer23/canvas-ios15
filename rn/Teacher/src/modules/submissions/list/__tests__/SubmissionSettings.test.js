//
// Copyright (C) 2017-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import React from 'react'
import { SubmissionSettings, mapStateToProps } from '../SubmissionSettings'
import renderer from 'react-test-renderer'
import explore from '../../../../../test/helpers/explore'

const template = {
  ...require('../../../../__templates__/helm'),
  ...require('../../../../__templates__/assignments'),
  ...require('../../../../__templates__/quiz'),
  ...require('../../../../redux/__templates__/app-state'),
}

let defaultProps = {
  courseID: '1',
  assignmentID: '2',
  navigator: template.navigator(),
  updateAssignment: jest.fn(),
  muted: false,
  assignment: template.assignment({ id: '2' }),
}

describe('SubmissionSettings', () => {
  beforeEach(() => jest.resetAllMocks())

  it('renders properly', () => {
    let tree = renderer.create(
      <SubmissionSettings {...defaultProps} />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('calls updateAssignment when mute toggle is pressed', () => {
    let tree = renderer.create(
      <SubmissionSettings {...defaultProps} />
    )

    let toggle = explore(tree.toJSON()).selectByID('submission-settings.muted') || {}
    toggle.props.onValueChange(true)

    expect(defaultProps.updateAssignment).toHaveBeenCalledWith(
      '1',
      { ...defaultProps.assignment, muted: true },
      defaultProps.assignment
    )
  })
})

describe('mapStateToProps', () => {
  let ownProps = {
    courseID: '1',
    assignmentID: '1',
  }

  it('returns the muted value of the assignment', () => {
    let state = template.appState({
      entities: {
        assignments: {
          '1': {
            data: template.assignment({ muted: true }),
          },
        },
      },
    })

    let props = mapStateToProps(state, ownProps)
    expect(props.muted).toEqual(true)
  })

  it('returns the assignment', () => {
    let state = template.appState({
      entities: {
        assignments: {
          '1': {
            data: template.assignment(),
          },
        },
      },
    })

    let props = mapStateToProps(state, ownProps)
    expect(props.assignment).toEqual(state.entities.assignments['1'].data)
  })
})
