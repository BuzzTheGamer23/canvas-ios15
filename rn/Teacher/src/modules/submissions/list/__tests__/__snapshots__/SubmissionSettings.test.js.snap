// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmissionSettings renders properly 1`] = `
<View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel="Mute Grades, Off"
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        Array [
          Object {
            "height": 60,
          },
          undefined,
          Object {
            "borderBottomColor": "#C7CDD1",
            "borderBottomWidth": 0.5,
          },
        ],
      ]
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "flex": 1,
            "flexDirection": "row",
            "paddingBottom": 12.8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 12.8,
          },
          Object {
            "backgroundColor": "white",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          numberOfLines={0}
          style={
            Array [
              Object {
                "fontFamily": ".SFUIDisplay",
              },
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
              },
              Array [
                Object {
                  "fontWeight": "600",
                },
                undefined,
              ],
              Object {
                "fontFamily": ".SFUIDisplay-semibold",
              },
            ]
          }
        >
          Mute Grades
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 0,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "flex-end",
            }
          }
        >
          <RCTSwitch
            border="bottom"
            disabled={false}
            height={60}
            identifier="submission-settings.muted"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            onTintColor="#374A59"
            onValueChange={[Function]}
            style={
              Array [
                Object {
                  "height": 31,
                  "width": 51,
                },
                undefined,
              ]
            }
            testID="submission-settings.muted"
            tintColor="#374A59"
            title="Mute Grades"
            value={false}
          />
        </View>
      </View>
    </View>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel="Anonymous Grading, Off"
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
        },
        Array [
          Object {
            "height": 60,
          },
          undefined,
          Object {
            "borderBottomColor": "#C7CDD1",
            "borderBottomWidth": 0.5,
          },
        ],
      ]
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "flex": 1,
            "flexDirection": "row",
            "paddingBottom": 12.8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 12.8,
          },
          Object {
            "backgroundColor": "white",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          numberOfLines={0}
          style={
            Array [
              Object {
                "fontFamily": ".SFUIDisplay",
              },
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
              },
              Array [
                Object {
                  "fontWeight": "600",
                },
                undefined,
              ],
              Object {
                "fontFamily": ".SFUIDisplay-semibold",
              },
            ]
          }
        >
          Anonymous Grading
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 0,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "flex-end",
            }
          }
        >
          <RCTSwitch
            border="bottom"
            disabled={false}
            height={60}
            identifier="submission-settings.anonymous"
            onChange={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            onTintColor="#374A59"
            onValueChange={[Function]}
            style={
              Array [
                Object {
                  "height": 31,
                  "width": 51,
                },
                undefined,
              ]
            }
            testID="submission-settings.anonymous"
            tintColor="#374A59"
            title="Anonymous Grading"
            value={false}
          />
        </View>
      </View>
    </View>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "fontFamily": ".SFUIDisplay",
        },
        Object {
          "color": "#7F91A7",
          "fontSize": 14,
        },
        Object {
          "paddingHorizontal": 12,
          "paddingVertical": 4,
        },
      ]
    }
  >
    This will anonymize each student and shuffle the submission list.
  </Text>
</View>
`;
